syntax = "proto3";

package synapse;

import "google/protobuf/empty.proto";
import "api/node.proto";
import "api/query.proto";
import "api/status.proto";

message Peripheral {
  string name = 1;
  string vendor = 2;
  uint32 peripheral_id = 3;
  NodeOptions options = 4;
}

message DeviceInfo {
  string name = 1;
  string serial = 2;
  uint32 synapse_version = 3;
  uint32 firmware_version = 4;
  Status status = 5;
  repeated Peripheral peripherals = 6;
  DeviceConfiguration configuration = 7;
}

message DeviceConfiguration {
  repeated NodeConfig nodes = 1;
  repeated NodeConnection connections = 2;
}

service SynapseDevice {
  rpc Info(google.protobuf.Empty) returns (DeviceInfo) {}
  rpc Configure(DeviceConfiguration) returns (Status) {}
  rpc Start(google.protobuf.Empty) returns (Status) {}
  rpc Stop(google.protobuf.Empty) returns (Status) {}
  rpc Query(QueryRequest) returns (QueryResponse) {}
}